nominal/core/datasource.py

```diff
@@ -0,0 +1,361 @@
+            exact_match: Filter the returned channels to those whose names match all provided strings
+                (case insensitive).
+            fuzzy_search_text: Filters the returned channels to those whose names fuzzily match the provided string.
+            tags: Dictionary of tags to filter channels by
+            start: The minimum data updated time to filter channels by
+            end: The maximum data start time to filter channels by
+
+        Yields:
+            Channel objects for each matching channel
+        """
+        next_page_token = None
+        while True:
+            query = datasource_api.SearchChannelsRequest(
+                data_sources=[self.rid],
+                exact_match=list(exact_match),
+                fuzzy_search_text=fuzzy_search_text,
+                previously_selected_channels={},
+                next_page_token=next_page_token,
+                page_size=None,
+                prefix=None,
+            )
+            response = self._clients.datasource.search_channels(self._clients.auth_header, query)
+            for channel_metadata in response.results:
+                yield Channel._from_conjure_datasource_api(self._clients, channel_metadata)
+            if response.next_page_token is None:
+                break
+            next_page_token = response.next_page_token
+
+    def to_pandas(
+        self,
+        channel_exact_match: Sequence[str] = (),
+        channel_fuzzy_search_text: str = "",
+        start: str | datetime | IntegralNanosecondsUTC | None = None,
+        end: str | datetime | IntegralNanosecondsUTC | None = None,
+        tags: dict[str, str] | None = None,
+    ) -> pd.DataFrame:
+        """Download a dataset to a pandas dataframe, optionally filtering for only specific channels of the dataset.
+
+        Args:
+        ----
+            channel_exact_match: Filter the returned channels to those whose names match all provided strings
+                (case insensitive).
+                For example, a channel named 'engine_turbine_rpm' would match against ['engine', 'turbine', 'rpm'],
+                whereas a channel named 'engine_turbine_flowrate' would not!
+            channel_fuzzy_search_text: Filters the returned channels to those whose names fuzzily match the provided
+                string.
+            tags: Dictionary of tags to filter channels by
+            start: The minimum data updated time to filter channels by
+            end: The maximum data start time to filter channels by
+
+        Returns:
+        -------
+            A pandas dataframe whose index is the timestamp of the data, and column names match those of the selected
+                channels.
+
+        Example:
+        -------
+        ```
+        import nominal as nm
+
+        rid = "..." # Taken from the UI or via the SDK
+        dataset = nm.get_dataset(rid)
+        s = dataset.to_pandas()
+        print("index:", s.index, "index mean:", s.index.mean())
```

Comment:
should use `df` rather than `s` (probably copied over from the channel one where the result is a `pd.Series` rather than a `pd.DataFrame`). Also `index` and `index.mean()` seem like kinda weird examples